Python/Product/Analysis/LanguageServer/EditorFiles.cs

```diff
@@ -0,0 +1,257 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.PythonTools.Analysis.Analyzer;
+using Microsoft.PythonTools.Intellisense;
+
+namespace Microsoft.PythonTools.Analysis.LanguageServer {
+    internal sealed class EditorFiles {
+        private readonly ConcurrentDictionary<Uri, EditorFile> _files = new ConcurrentDictionary<Uri, EditorFile>();
+        private readonly Server _server;
+
+        public EditorFiles(Server server) {
+            _server = server;
+        }
+
+        public EditorFile GetDocument(Uri uri) => _files.GetOrAdd(uri, _ => new EditorFile(_server));
+        public void Remove(Uri uri) => _files.TryRemove(uri, out _);
+        public void Open(Uri uri) => GetDocument(uri).Open(uri);
+        public void Close(Uri uri) => GetDocument(uri).Close(uri);
+
+        public void UpdateDiagnostics() {
+            foreach (var entry in _server.ProjectFiles.All) {
```

Comment:
`_server.ProjectFiles.All` returns values from `ConcurrentDictionary`, so it can be modified while enumerator is working. Do we care about that?

Reply:
If file is gone, we do want to skip it so there is no harm. If file was added and diagnostic is available, then good too.